SpotData CLI and framework.

bayliffe opened the pull request:
A component of #34 

Remaining framework files. Now passes tests as other components have been merged.

Testing:
 - [ ] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

benfitzpatrick commented:
Please make the command a bit more 'action-ey' - `improver spot-extract` sounds slightly dermatological, but something like that. Could you also rename `framework` to something else? `main`?

bayliffe commented:
Here is an example of the diagnostic config json file:

```
{
    "high_cloud_amount": {
        "diagnostic_name": "high_type_cloud_area_fraction",
        "extrema": false,
        "filepath": "high_cloud_amount",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "low_cloud_amount": {
        "diagnostic_name": "low_type_cloud_area_fraction",
        "extrema": false,
        "filepath": "low_cloud_amount",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "medium_cloud_amount": {
        "diagnostic_name": "medium_type_cloud_area_fraction",
        "extrema": false,
        "filepath": "medium_cloud_amount",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "relative_humidity": {
        "diagnostic_name": "relative_humidity",
        "extrema": false,
        "filepath": "relative_humidity_at_screen_level",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "surface_pressure": {
        "diagnostic_name": "surface_air_pressure",
        "extrema": false,
        "filepath": "surface_pressure",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "temperature": {
        "diagnostic_name": "air_temperature",
        "extrema": true,
        "filepath": "temperature_at_screen_level",
        "interpolation_method": "model_level_temperature_lapse_rate",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "temperature_orog": {
        "diagnostic_name": "air_temperature",
        "extrema": true,
        "filepath": "temperature_at_screen_level",
        "interpolation_method": "orography_derived_temperature_lapse_rate",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "total_cloud_amount": {
        "diagnostic_name": "cloud_area_fraction",
        "extrema": false,
        "filepath": "total_cloud",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "visibility": {
        "diagnostic_name": "visibility_in_air",
        "extrema": true,
        "filepath": "visibility_at_screen_level",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "wind_direction": {
        "diagnostic_name": "wind_from_direction",
        "extrema": false,
        "filepath": "horizontal_wind_speed_and_direction_at_10m",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    },
    "wind_speed": {
        "diagnostic_name": "wind_speed",
        "extrema": false,
        "filepath": "horizontal_wind_speed_and_direction_at_10m",
        "interpolation_method": "use_nearest",
        "neighbour_finding": {
            "land_constraint": false,
            "method": "fast_nearest_neighbour",
            "vertical_bias": null
        }
    }
}
```

bayliffe commented:
Here is an example of a constants json file:

```
{
    "dthetadz_threshold": 0.02,
    "dz_max_adjustment": 70,
    "dz_tolerance": 2.0
}
```

bayliffe commented:
Here is an example of a site definition json file:
```
{"altitude": 0, "bestdata_id": 1, "gmtoffset": -3.0, "latitude": -51.0, "longitude": -55.5},
{"altitude": 205, "bestdata_id": 2, "gmtoffset": -4.0, "latitude": -47.5, "longitude": -75}
```

benfitzpatrick commented:
I'd still prefer 'utcoffset' to 'gmtoffset', but other than that the JSON files look good to me.

meabhnicguidhir commented on lib/improver/tests/spotdata/spotdata/test_framework.py:
```diff
@@ -0,0 +1,278 @@
+        Call(['rmdir', self.configs])
+        Call(['rmdir', self.output_directory])
+
+
```
This is an awful lot of setup to test two exceptions are being raised. For example, do you need the `additional_data` setup? When testing just these two errors will you ever reach that stage of the code?

bayliffe replied:
Agreed. I've simplified to be only the necessary stuff to cover the three tests. I chose a simpler method of extracting data, so that removed the need for all the multi-level data.

meabhnicguidhir commented on lib/improver/tests/spotdata/spotdata/test_framework.py:
```diff
@@ -0,0 +1,278 @@
+        with self.assertRaisesRegexp(ValueError, msg):
+            Plugin(*self.args, **self.kwargs)
+
+
```
Could you add a test for the `Exception` at the end of `process_diagnostic` (see comment in `framework.py`).

bayliffe replied:
:white_check_mark:

meabhnicguidhir commented on lib/improver/tests/spotdata/spotdata/test_times.py:
```diff
@@ -0,0 +1,132 @@
+from datetime import timedelta
+from datetime import time
+
+from improver.spotdata.times import get_forecast_times as Plugin
```
Again, could `Plugin` be changed to `Function` or something which more accurately reflects what get_forecast_rimes` is?

Also, change the test titles below to match the style guide.

bayliffe replied:
Plugin changed to function.
As far as I can see the titles are in the format being used elsewhere.

meabhnicguidhir commented on lib/improver/tests/spotdata/spotdata/test_times.py:
```diff
@@ -0,0 +1,132 @@
+        forecast_start = dt(2017, 6, 1, 9, 0)
+        forecast_date = forecast_start.strftime("%Y%m%d")
+        forecast_time = int(forecast_start.strftime("%H"))
+        forecast_length = 144
```
Maybe set this to something other than the default just to be clear it is using a different length?

bayliffe replied:
:white_check_mark:

meabhnicguidhir commented on lib/improver/tests/spotdata/spotdata/test_framework.py:
```diff
@@ -0,0 +1,278 @@
+class Test_missing_data(Test_run_framework):
+    """
+    Test the run_framework interface with various options.
+
```
Change the test titles below to match the style guide.

bayliffe replied:
Done, I think, still waiting on a style guide for unit tests.

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+
+from improver.spotdata.read_input import (Load,
+                                          get_method_prerequisites,
+                                          read_config)
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+                                                extract_cube_at_time,
+                                                extract_ad_at_time)
+
+
```
No. It was all written rather standalone. It probably needs to be restructured in a more plugin style way that can be controlled from the suite. I'm still waiting on a better understanding of how exactly it's going to be used before I start rearranging it. I have a meeting on Friday that may clarify things, but for now it remains a bit of an improver oddity.

benfitzpatrick replied:
Should we really put it in if we know it will need serious refactoring?

benfitzpatrick replied:
I think yes, but we will need to commit to prioritising actually doing the refactoring

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+        tolerances that may be set. e.g. maxiumum vertical extrapolation/
+        interpolation of temperatures using a temperature lapse rate method.
+
+    site_properties : dict
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+
+    Returns:
+    --------
+    Writes out cubes of the requested diagnostics, with data extracted to the
```
For want of a better name at present and pending a different method of using this, I'll leave it as returns.

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+        config_constants = read_config(constants_file_path)
+
+    # Use the diagnostic_list to estabilish which diagnostics are to be
+    # processed; if unset, use all.
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+        neighbour_hash = construct_neighbour_hash(neighbour_finding)
+        # Check if defined neighbour method results already exist.
+        if neighbour_hash not in neighbours.keys():
+            print 'Making neighbours ', neighbour_hash
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+                )
+
+    if use_multiprocessing:
+        # Process diagnostics on separate threads is multiprocessing is
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+    Returns:
+    --------
+    Nil.
+
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+    neighbour_hash = construct_neighbour_hash(diagnostic['neighbour_finding'])
+    neighbour_list = neighbours[neighbour_hash]
+
+    # Check to see if additional diagnostics are needed (e.g. multi-level data)
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/framework.py:
```diff
@@ -0,0 +1,335 @@
+        cube_out, = resulting_cubes.concatenate()
+        WriteOutput('as_netcdf').process(cube_out, output_path)
+    else:
+        raise Exception('No data extracted.')
```
Correct. Test added.

bayliffe commented on bin/improver-spotdata:
```diff
@@ -0,0 +1,132 @@
+                  forecast_length=args.forecast_length,
+                  output_path=args.output_path,
+                  use_multiprocessing=args.multiprocess)
+
```
Ultimately I think that 'framework.py' (now main.py) should be divided between the CLI and the suite, but the suite doesn't really exist in a fully functional form yet.

benfitzpatrick replied:
I agree with Meabh - but OK, I think we should revisit this later then.

benfitzpatrick commented:
Still have a comment on the naming: can you think of a better name than `improver spot-data`? `improver site-extract`?

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,89 @@
+    times. This list will be created from input specifications if provided.
+    Otherwise defaults are to start today at the most recent 6-hourly interval
+    (00, 06, 12, 18) and to run out to T+144 hours.
+
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,89 @@
+
+    forecast_date : string (YYYYMMDD)
+        A string of format YYYYMMDD defining the start date for which forecasts
+        are required.
```
:white_check_mark:

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,89 @@
+
+    forecast_time : integer
+        An integer giving the hour on the forecast_date at which to start the
+        forecast output; 24hr clock such that 17 = 17Z for example.
```
:white

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,89 @@
+    forecast_length : integer
+        An integer giving the desired length of the forecast output in hours
+        (e.g. 48 for a two day forecast period).
+
```
This should not be hard coded. I envisage this will be a suite variable that can be pulled in, but it doesn't exist yet.

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,89 @@
+                      range(min(forecast_length, 49))]
+    forecast_times = (forecast_times +
+                      [forecast_start_time + timedelta(hours=x) for x in
+                       range(51, min(forecast_length+1, 144+1), 3)])
```
Agreed, and I assume they'll be another suite variable.

bayliffe commented on lib/improver/tests/spotdata/spotdata/test_framework.py:
```diff
@@ -0,0 +1,278 @@
+from collections import OrderedDict
+import numpy as np
+
+from improver.spotdata.framework import run_framework as Plugin
```
Okay, why not.

bayliffe commented:
@benfitzpatrick I have renamed the CLI to improver-spot-extract despite its slightly dermatological sound. I have renamed framework to main and will be keen to get this stuff into the suite when it's possible.

benfitzpatrick commented on lib/improver/tests/spotdata/spotdata/test_main.py:
```diff
@@ -0,0 +1,201 @@
+
+    def tearDown(self):
+        """Remove temporary directories created for testing."""
+        Call(['rm', '-f', self.data_path])
```
There are `os` library ways of doing this

gavinevans replied:
Yes, either `os.rmdir` or `shutil.rmtree` might suit your purpose.

bayliffe replied:
Changed to use shutil,rmtree.

gavinevans commented on lib/improver/tests/spotdata/spotdata/test_main.py:
```diff
@@ -0,0 +1,196 @@
+        iris.save(cube, self.data_path)
+        iris.save(orography, self.orography_path)
+
+        self.config_path = self.data_directory + '/spotdata_constants.json'
```
It might be worth adding a few inline comments to this `setUp` method to help highlight what you're trying to set up.

bayliffe replied:
Done.

gavinevans commented on lib/improver/spotdata/main.py:
```diff
@@ -0,0 +1,361 @@
+    # Concatenate CubeList into Cube, creating a time DimCoord, and write out.
+    if resulting_cubes:
+        cube_out, = resulting_cubes.concatenate()
+        WriteOutput('as_netcdf', dir_path=output_path).process(cube_out)
```
I'm slightly concerned about the value of the `WriteOutput` plugin, as essentially it's an elaborate wrapper for:
```python
iris.save(cube, '{}/{}.nc'.format(self.dir_path, cube.name()))
```
I'm a bit confused about the value of the `as_netcdf` method. Are you anticipating that you'll need to write out spot data output into a file format that Iris doesn't support? If you actually returned cubes to your higher level functions and saved at the top level, then it would make it easier to test for a successful run within e.g. `process_diagnostic` and `run_spotdata`.

bayliffe replied:
Once there is a suite I will be happy to change main into part of the suite. Currently cubes are written out at more than one point (data extraction and extrema extraction).

bayliffe replied:
I am not sure whether we will have to do something slightly different for spotdata going to verification, possibly pushing something to csv or a numpy array for pandas to push into a SQL database, The wrapper is basically there so I can add additional options and it can be called to give verification a separate stream in necessary. Does that make sense?

gavinevans replied:
Yes, I thought that might be your reasoning. I think it might make sense to put this saving functionality into a top-level utility, so there might be technical debt issue to do this reorganisation.

gavinevans commented on lib/improver/tests/spotdata/spotdata/test_main.py:
```diff
@@ -0,0 +1,196 @@
+        Call(['rmdir', self.output_directory])
+
+
+class Test_run_spotdata(Test_main):
```
Can you add a test to show that this has been successful? I can see that this is a bit difficult because nothing is returned, as the file is written down in `WriteOutput`. From a testing perspective, it's a shame that the cubes aren't passed back up and the output isn't written out at a higher level, ideally in a spot data CLI.

bayliffe replied:
The writing routine is tested in its own unit tests, including a test that output is produced.

gavinevans commented on lib/improver/tests/spotdata/spotdata/test_main.py:
```diff
@@ -0,0 +1,196 @@
+class Test_process_diagnostic(Test_main):
+    """Test the process_diagnostic function."""
+
+    def test_no_data_files(self):
```
Can you add a successful test?

bayliffe replied:
Added.

bayliffe commented on bin/improver-spot-extract:
```diff
@@ -0,0 +1,148 @@
+
+
+def valid_latitude(value):
+    """Ensures latitude values fall in allowed range."""
```
Done.

bayliffe commented on bin/improver-spot-extract:
```diff
@@ -0,0 +1,148 @@
+
+
+def valid_longitude(value):
+    """Ensures longitude values fall in allowed range."""
```
Done.

bayliffe commented on bin/improver-spot-extract:
```diff
@@ -0,0 +1,148 @@
+    return value
+
+def site_dictionary(latitudes, longitudes, altitudes):
+    if len(latitudes) != len(longitudes):
```
Done.

bayliffe commented on bin/improver-spot-extract:
```diff
@@ -0,0 +1,148 @@
+        site_properties = site_dictionary(args.latitudes, args.longitudes,
+                                          args.altitudes)
+
+    print site_properties
```
Removed.

bayliffe commented on lib/improver/spotdata/extract_data.py:
```diff
@@ -207,28 +227,41 @@ def make_cube(self, cube, data, sites):
         """
-        time_coord = cube.coord('time')
-        time_coord.convert_units('hours since 1970-01-01 00:00:00')
+        # Ensure leading dimension axis is time.
```
I've changed this. I only need to ensure that time is a dimension coordinate, it needn't be leading.

bayliffe commented on lib/improver/spotdata/extract_data.py:
```diff
@@ -207,28 +227,41 @@ def make_cube(self, cube, data, sites):
+
+        n_non_spatial_dimcoords = len(cube.dim_coords) - 2
+        non_spatial_dimcoords = cube.dim_coords[0:n_non_spatial_dimcoords]
+        for coord in non_spatial_dimcoords:
```
Oh yes, having a silly turn.

bayliffe commented on lib/improver/spotdata/main.py:
```diff
@@ -0,0 +1,361 @@
+
+    neighbours : numpy.array
+        Array of neigbouring grid points that are associated with sites
+        in the SortedDictionary of sites.
```
No, it's sorted, it's really got itself together lately. Changed.

bayliffe commented on lib/improver/spotdata/main.py:
```diff
@@ -0,0 +1,361 @@
+        Array of neigbouring grid points that are associated with sites
+        in the SortedDictionary of sites.
+
+    sites : dict
```
Done.

bayliffe commented on lib/improver/spotdata/main.py:
```diff
@@ -0,0 +1,361 @@
+
+    Returns:
+    --------
+    Nil.
```
Happy to change it.

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,102 @@
+        forecast_start_time = dt.combine(
+            start_date, time(divmod(dt.utcnow().hour, 6)[0]*6))
+
+    if forecast_length is None:
```
I've made it an arg, and set the default in main.py. This removes the need to check the type or enforce an assumed length in here.

bayliffe commented on lib/improver/spotdata/times.py:
```diff
@@ -0,0 +1,102 @@
+    if forecast_length is None:
+        forecast_length = 144
+
+    # Generate forecast times. Hourly to T+48, 3 hourly to T+forecast_length.
```
I've actually changed my mind about changing this. It's horrible, agreed. For now it does what we want. I will start writing a ticket about moving this stuff into the suite, and I'll make sure to include something about generalising this. Is that okay for now?

gavinevans replied:
I think this is satisfactory for now, although yes, this does need to be documented within an issue to highlight this technical debt. If you put the details of the problem on the issue, then you or @benfitzpatrick can create a technical debt issue.

bayliffe commented:
Much of the work in this ticket needs to be revisited once a suite is up and running. Many of the tasks here would be better connected using a suite, and many of the variables could do with being set at the suite level (e.g. forecast times, output frequency (hourly to T+48, three-hourly to T+168), etc.).

bayliffe commented:
A ticket has been created to tackle the structural issues caused by the need for SpotData to be ready prior to the suite being finished, #217. This will need to be addressed once the multi-model suite is written.
